pallets/crowdloan/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(test)]
+pub(crate) mod mock;
+
+pub use crate::pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{Currency, StorageVersion},
+	};
+	use frame_system::{pallet_prelude::*, EnsureOneOf, EnsureRoot, EnsureSigned, WeightInfo};
+	use sp_runtime::{
+		traits::{BadOrigin, StaticLookup},
+		Either,
+	};
+
+	pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+	pub type BalanceOf<T> = <<T as Config>::Currency as Currency<AccountIdOf<T>>>::Balance;
+
+	pub const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Currency type.
+		type Currency: Currency<AccountIdOf<Self>>;
+		/// Overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub admin_account: AccountIdOf<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				admin_account: AccountIdOf::<T>::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			AdminAccount::<T>::set(self.admin_account.clone());
+		}
+	}
+
+	/// The administrator account allowed to manage the pallet storage.
+	#[pallet::storage]
+	#[pallet::getter(fn admin_account)]
+	pub type AdminAccount<T> = StorageValue<_, AccountIdOf<T>, ValueQuery>;
+
+	/// The set of contributions.
+	///
+	/// It maps from contributor's account to amount contributed.
+	#[pallet::storage]
+	#[pallet::getter(fn contributions)]
+	pub type Contributions<T> = StorageMap<_, Blake2_128Concat, AccountIdOf<T>, BalanceOf<T>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A new admin has been set.
+		/// \[old administrator account, new administrator account\]
+		NewAdminAccountSet(AccountIdOf<T>, AccountIdOf<T>),
+		/// A new contribution has been set.
+		/// \[contributor account, old amount (OPTIONAL), new amount\]
+		NewContributionSet(AccountIdOf<T>, Option<BalanceOf<T>>, BalanceOf<T>),
+		/// A contribution has been removed.
+		/// \[contributor account\]
+		ContributionRemoved(AccountIdOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The to-delete contribution is not present.
+		ContributorNotPresent,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Sets a new account as the admin of this pallet.
+		///
+		/// The dispatch origin can be either Sudo or the current admin account.
+		///
```

Comment:
M: Missing the emitted event.

```suggestion
		///
		/// Emits `NewAdminAccountSet `.
		///
```